library/data/component/src/main/java/org/quiltmc/qsl/component/impl/container/LazifiedComponentContainer.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.component.impl.container;
+
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.util.Identifier;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.component.api.Component;
+import org.quiltmc.qsl.component.api.ComponentContainer;
+import org.quiltmc.qsl.component.api.ComponentProvider;
+import org.quiltmc.qsl.component.api.ComponentType;
+import org.quiltmc.qsl.component.api.components.NbtComponent;
+import org.quiltmc.qsl.component.api.components.SyncedComponent;
+import org.quiltmc.qsl.component.api.components.TickingComponent;
+import org.quiltmc.qsl.component.api.event.ComponentEvents;
+import org.quiltmc.qsl.component.impl.ComponentsImpl;
+import org.quiltmc.qsl.component.impl.sync.header.SyncPacketHeader;
+import org.quiltmc.qsl.component.impl.sync.packet.SyncPacket;
+import org.quiltmc.qsl.component.impl.sync.packet.PacketIds;
+import org.quiltmc.qsl.component.impl.util.ErrorUtil;
+import org.quiltmc.qsl.component.impl.util.Lazy;
+import org.quiltmc.qsl.component.impl.util.StringConstants;
+import org.quiltmc.qsl.networking.api.ServerPlayNetworking;
+
+import java.util.*;
+import java.util.function.Supplier;
+
+public class LazifiedComponentContainer implements ComponentContainer {
```

Comment:
Possibly rename this to `LazyComponentContainer`? "Lazified" makes no sense to me.

Reply:
This was also a bad use of ArrayList. I should've used a Set since the only method called on this collection is 'contains'.